clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -980,6 +1002,28 @@ void handleCoordinatorReady() {
                 null;
         this.coordinatorSupportsBumpingEpoch = initProducerIdVersion != null &&
                 initProducerIdVersion.maxVersion() >= 3;
+
+        if (nodeApiVersions == null) return;
+
+        if (nodeApiVersions.finalizedFeatures() != null) {
+            /*
+                To enable the transaction V2, it requires:
+                1. transaction.version finalized version >= 2
```

Comment:
We should check every node API versions (not just coordinator's) and keep track of the latest transaction version (where "latest" is defined as the latest epoch).  So that logic would be something like this:

1. When we get API versions from any node, check if the transaction manager has the epoch that is less than the new epoch.
2. If the epoch stored in the transaction manager is less than the new epoch, store the new value and the new epoch in the transaction manager.
3. Have a method `boolean coordinatorSupportsTransactionV2() { return this.transactionVersion >= 2; }`

This way we learn the cluster-wide consensus on the transaction version, vs. individual brokers that may have different value so our view doesn't jump back and forth (if coordinator moves, for example).

Reply:
This is a good point. I agree 👍

Reply:
> When we get API versions from any node

The optimal way of handling this is to have an ApiVersionsRequest callback and check the transaction version in it. However, it is a bit tricky here. The ApiVersionsRequest is handled in the KafkaClient, a layer below the Sender class. It does not expose any interaction method other than the _apiVersions_ object. If we want to register a callback, we should have an interface change in the KafkaClient which may need a KIP?
A suboptimal way is to check the _apiVersions_ for all the node features when calling the _coordinatorSupportsTransactionV2_. I guess it is acceptable as the checking does not cost much.

Reply:
Maybe I'm missing something, but can we store this within ApiVersions and update the value every time a new apiVersions request is handled? We already call  `apiVersions.update(node, nodeVersionInfo);` and that does a calculation for max produce magic, so could we do another method in update to check the epoch and if it is greater than the existing epoch, we update the internal value in ApiVersions? (We may need to add epoch to NodeApiVersions -- and this is not a public api)

Reply:
I did something a bit different here. Every time the ApiVersions are updated, the maxFinalizedFeaturesEpoch among the NodeApiVersions is stored. Then the transaction manager can just check this maxFinalizedFeaturesEpoch to determine if it needs to check the feature updates.

Reply:
I think it gets the same think accomplished. 👍